### PR TITLE
feat: [CDS-44562]: Instance Sync V2 for K8s in CG

### PR DESCRIPTION
**Pull request by shubham-in10se**
## Describe your changes
More details about the design are present in this doc: https://harness.atlassian.net/wiki/spaces/CDP/pages/21203058735/Instance+Sync+Architecture+V2

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/resources/InstanceSyncResource.java [hunk 1]
```diff
@@ -78,4 +86,37 @@ public RestResponse<Boolean> processInstanceSyncNGResult(
...
+      log.error("Failed to process results for perpetual task: [{}]", perpetualTaskId.replaceAll("[\r\n]", ""), e);
+    }
+    return new RestResponse<>(true);
+  }
+
+  @DelegateAuth
+  @GET
+  @Path("instance-sync-v2/task-details/{perpetualTaskId}")
+  @Produces(ProtocolBufferMediaType.APPLICATION_PROTOBUF)
+  public Response processInstanceSyncV2Result(
```
Comment (shubham-in10se) [inline 400-rest/src/main/java/software/wings/resources/InstanceSyncResource.java:110]: rename this method

**Comment by shubham-in10se** [comment]
trigger ti0

**Comment by shubham-in10se** [comment]
trigger ti1

**Comment by shubham-in10se** [comment]
trigger ti2

**Comment by shubham-in10se** [comment]
trigger ti3

**Comment by shubham-in10se** [comment]
trigger ti4

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 37 of 37 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: :shipit: complete! ✓

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 33 of 37 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: :shipit: complete! ✓

___
*[`260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/CgK8sInstancesDetailsFetcher.java` line 81 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwXFHX9ga3rXSdjPtl:-NFwXFHX9ga3rXSdjPtm:b7hkvsl) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/CgK8sInstancesDetailsFetcher.java#L81)):*
> ```Java
>           .setTaskDetailsId(releaseDetails.getTaskDetailsId())
>           .setExecutionStatus(
>               runningK8sPods != null ? CommandExecutionStatus.SUCCESS.name() : CommandExecutionStatus.FAILURE.name())
> ```

Will this ever be null? or will it be at least an empty list
___
*[`260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/InstanceDetailsFetcherFactory.java` line 20 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwVtEu4A4HgHcGABGz:-NFwVtEu4A4HgHcGABH-:bdpy2q7) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/InstanceDetailsFetcherFactory.java#L20)):*
> ```Java
> public class InstanceDetailsFetcherFactory {
>   private CgK8sInstancesDetailsFetcher k8sInstancesDetailsFetcher;
>   private ConcurrentHashMap<String, InstanceDetailsFetcher> holder;
> ```

nit: These could be final



_Code quote:_
```Java
  private CgK8sInstancesDetailsFetcher k8sInstancesDetailsFetcher;
  private ConcurrentHashMap<String, InstanceDetailsFetcher> holder;
```
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/handler/CgInstanceSyncV2HandlerFactory.java` line 20 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwoE2j7cD8e0fxiJvM:-NFwoE2j7cD8e0fxiJvN:b-n2rh3b) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/handler/CgInstanceSyncV2HandlerFactory.java#L20)):*
> ```Java
>   private K8sInstanceSyncV2HandlerCg k8sHandler;
> 
>   private ConcurrentHashMap<SettingVariableTypes, CgInstanceSyncV2Handler> holder;
> ```

nit: these fields can be final



_Code quote:_
```Java
  private K8sInstanceSyncV2HandlerCg k8sHandler;

  private ConcurrentHashMap<SettingVariableTypes, CgInstanceSyncV2Handler> holder;
```
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java` line 155 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFx6cz3-xx2lX5M4PI2:-NFx6cz3-xx2lX5M4PI3:b895e60) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java#L155)):*
> ```Java
>                 .filter(existingIdentifier -> existingIdentifier instanceof CgK8sReleaseIdentifier)
>                 .filter(existingIdentifier
>                     -> StringUtils.equals(((CgK8sReleaseIdentifier) existingIdentifier).getReleaseName(),
> ```

Is it possible to make this more readable? :thinking: 
Maybe we could do .map(CgK8sReleaseIdentifier.class::cast) and change the return type to Optional<CgK8sReleaseIdentifier>?
newIdentifier also could be Type cast at the beginning itself?
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/service/CgInstanceSyncTaskDetailsService.java` line 55 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFx7hFnBVZn-PG341gE:-NFx7hFnBVZn-PG341gF:btfwube) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/service/CgInstanceSyncTaskDetailsService.java#L55)):*
> ```Java
>     Query<InstanceSyncTaskDetails> query = mongoPersistence.createQuery(InstanceSyncTaskDetails.class)
>                                                .filter(InstanceSyncTaskDetailsKeys.accountId, accountId)
>                                                .filter(InstanceSyncTaskDetailsKeys.cloudProviderId, cloudProviderId);
> ```

Do we need an index covering cloudProviderId?

**Review by shubham-in10se — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: 10 file(s) to review. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/CgK8sInstancesDetailsFetcher.java` line 81 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwXFHX9ga3rXSdjPtl:-NFy-Vt98q77pZ7qnTMt:b6me11i) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/CgK8sInstancesDetailsFetcher.java#L81)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Will this ever be null? or will it be at least an empty list
</blockquote></details>

if it is empty list, then we consider the run to be a success, it only in case it is returned as `null` do we mark it as failure.
___
*[`260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/InstanceDetailsFetcherFactory.java` line 20 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwVtEu4A4HgHcGABGz:-NFy0HYm8vt7lh2OGK33:b-7rl0kd) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/InstanceDetailsFetcherFactory.java#L20)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

nit: These could be final
</blockquote></details>

done
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/handler/CgInstanceSyncV2HandlerFactory.java` line 20 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwoE2j7cD8e0fxiJvM:-NFy0OZJ7sZZ3rW6hB48:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/handler/CgInstanceSyncV2HandlerFactory.java#L20)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

nit: these fields can be final
</blockquote></details>

done
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java` line 155 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFx6cz3-xx2lX5M4PI2:-NFy1fs02ClfhVN9RRaT:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java#L155)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Is it possible to make this more readable? :thinking: 
Maybe we could do .map(CgK8sReleaseIdentifier.class::cast) and change the return type to Optional<CgK8sReleaseIdentifier>?
newIdentifier also could be Type cast at the beginning itself?
</blockquote></details>

done
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/service/CgInstanceSyncTaskDetailsService.java` line 55 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFx7hFnBVZn-PG341gE:-NFy1skHDJkvlhexSeVB:bft4x9q) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/service/CgInstanceSyncTaskDetailsService.java#L55)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Do we need an index covering cloudProviderId?
</blockquote></details>

yes.
we are making a call here: `CgInstanceSyncTaskDetailsService.fetchForCloudProvider()`
___
*[`400-rest/src/main/java/software/wings/resources/InstanceSyncResource.java` line 110 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFe9g6c9bqzgg3tI_MB-r1-110:-NFy27gGDG7xBoUzM4oZ:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/resources/InstanceSyncResource.java#L110)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

rename this method
</blockquote></details>

done

**Review by shubham-in10se — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: 10 file(s) to review. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/CgK8sInstancesDetailsFetcher.java` line 81 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFwXFHX9ga3rXSdjPtl:-NFyqtvd5__qDvJIYLZ_:byvd1ev) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/260-delegate/src/main/java/io/harness/perpetualtask/instancesyncv2/cg/CgK8sInstancesDetailsFetcher.java#L81)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

if it is empty list, then we consider the run to be a success, it only in case it is returned as `null` do we mark it as failure.
</blockquote></details>

nevertheless, its a check for safety. If in future, some fetcher returns null instead of emptyList.
___
*[`400-rest/src/main/java/software/wings/instancesyncv2/service/CgInstanceSyncTaskDetailsService.java` line 55 at r1](https://reviewable.io/reviews/harness/harness-core/39389#-NFx7hFnBVZn-PG341gE:-NFyr31L3HYNxCLIirDU:bjehtn4) ([raw file](https://github.com/harness/harness-core/blob/59cf34256d5c1ee3ccab64ba5104441e47f04d1e/400-rest/src/main/java/software/wings/instancesyncv2/service/CgInstanceSyncTaskDetailsService.java#L55)):*
<details><summary><i>Previously, shubham-in10se (Shubham Maheshwari) wrote…</i></summary><blockquote>

yes.
we are making a call here: `CgInstanceSyncTaskDetailsService.fetchForCloudProvider()`
</blockquote></details>

added the index.

**Review by vikyathharekal — COMMENTED** [review]
There are a few scenarios that could be covered by UT. Are we planning to add them later or in this PR itself?

> Reviewed 4 of 37 files at r1, 9 of 10 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389#-:-NG1-6OpDqOmcS1pneH5:b-y4qeil)* status: 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java` line 278 at r3](https://reviewable.io/reviews/harness/harness-core/39389#-NG11soVBCeXoTr-dYBc:-NG11soVBCeXoTr-dYBd:bty90k6) ([raw file](https://github.com/harness/harness-core/blob/bccb9fe756ccb5a2488080f2fb8ebac3fd201baa/400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java#L278)):*
> ```Java
>       return Collections.emptyList();
>     }
>     ContainerInfrastructureMapping containerInfraMapping = (ContainerInfrastructureMapping) infraMapping;
> ```

I guess we'll reach here for other non-ContainerInfrastructureMapping as well.

**Review by vikyathharekal — COMMENTED** [review]
nit: Some mocks are unused, could clean up if making another commit.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389#-:-NG13BKf0CI3LCc8qakV:bquzuns)* status: 1. unresolved discussions. 0 of 1 approvals obtained.

**Review by shubham-in10se — COMMENTED** [review]
more UTs will be added as we onboard more cases onto the new framework.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389#-:-NG18Oqe7IWJYpsrlxqq:b-5v3xlx)* status: 1 file(s) to review. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java` line 278 at r3](https://reviewable.io/reviews/harness/harness-core/39389#-NG11soVBCeXoTr-dYBc:-NG1826m0b8gr3fJ65Ri:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/bccb9fe756ccb5a2488080f2fb8ebac3fd201baa/400-rest/src/main/java/software/wings/instancesyncv2/handler/K8sInstanceSyncV2HandlerCg.java#L278)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

I guess we'll reach here for other non-ContainerInfrastructureMapping as well. 
</blockquote></details>

Done.

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: :shipit: complete! ✓

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 29 of 37 files at r1, 9 of 10 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: :shipit: complete! ✓

___
*[`400-rest/src/main/java/software/wings/instancesyncv2/InstanceSyncV2Module.java` line 12 at r2](https://reviewable.io/reviews/harness/harness-core/39389#-NFyaBVvAnxa5kJBmt4B:-NFyaBVvAnxa5kJBmt4C:b-87zl4m) ([raw file](https://github.com/harness/harness-core/blob/7a5a1d13df9061a9cc62b1ce63eaf9d64e6789f8/400-rest/src/main/java/software/wings/instancesyncv2/InstanceSyncV2Module.java#L12)):*
> ```Java
> import com.google.inject.AbstractModule;
> 
> public class InstanceSyncV2Module extends AbstractModule {
> ```

Is this module used/actually required?

**Review by shubham-in10se — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: 10 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/instancesyncv2/InstanceSyncV2Module.java` line 12 at r2](https://reviewable.io/reviews/harness/harness-core/39389#-NFyaBVvAnxa5kJBmt4B:-NG1kXMA5_zwqcKEsBCP:brpmqc3) ([raw file](https://github.com/harness/harness-core/blob/7a5a1d13df9061a9cc62b1ce63eaf9d64e6789f8/400-rest/src/main/java/software/wings/instancesyncv2/InstanceSyncV2Module.java#L12)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Is this module used/actually required?
</blockquote></details>

removed.

**Review by abosii-harness — APPROVED** [review]
> Reviewed 10 of 10 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: :shipit: complete! ✓

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 10 of 10 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39389)* status: :shipit: complete! ✓

**Pull request merged**